File: src/main/java/com/epri/dss/shared/impl/PointerListImpl.java
package com.epri.dss.shared.impl;

import com.epri.dss.shared.PointerList;

public class PointerListImpl implements PointerList {

	private int numInList;

	private int maxAllocated;

	private int activeItem;

	private Object[] list;

	private int incrementSize;

	public PointerListImpl(int size) {
		super();
		maxAllocated = size;
		// default size & increment
		if (maxAllocated <= 0) maxAllocated = 10;
		list = new Object[maxAllocated];
		numInList = 0;
		activeItem = 0;
		// increment is equal to original allocation
		incrementSize = maxAllocated;
	}

	public Object getFirst() {
		if (numInList > 0) {
		activeItem = 0;
		return list[activeItem];
		} else {
			activeItem = -1;
			return null;
		}
	}

	public Object getNext() {
		if (numInList > 0) {
		activeItem += 1;
		if (activeItem > numInList) {
			activeItem = numInList;
			return null;
		} else {
			return list[activeItem];
		}
		} else {
			activeItem = -1;
			return null;
		}
	}

	public Object getActive() {
		if ((activeItem > 0) && (activeItem <= numInList)) {
			return get(activeItem);
		} else {
			return null;
		}
	}

	public void setNew(Object value) {
		add(value);
	}

	public void clear() {
		activeItem = -1;
		numInList = 0;
	}

	/** Returns index of item */
	public int add(Object p) {
		numInList += 1;
		if (numInList > maxAllocated) {
			maxAllocated = maxAllocated + incrementSize;
			// FIXME: Resize array
			list = new Object[maxAllocated];
		}
		list[numInList] = p;
		activeItem = numInList;
		return numInList;
	}

	public Object get(int i) {
		if ((i < 1) || (i > numInList)) {
			return null;
		} else {
			activeItem = i;
			return list[i];
		}
	}

	public int size() {
		return numInList;
	}

	public int getActiveIndex() {
		return activeItem;
	}

}

File: pydss/shared/PointerListImpl.py
from dss.shared.PointerList import (PointerList,)


class PointerListImpl(object, PointerList):
    numInList = None
    maxAllocated = None
    activeItem = None
    list = None
    incrementSize = None

    def __init__(self, size):
        super(PointerListImpl, self)()
        self.maxAllocated = size
        # default size & increment
        if self.maxAllocated <= 0:
            self.maxAllocated = 10
        self.list = [None] * self.maxAllocated
        self.numInList = 0
        self.activeItem = 0
        # increment is equal to original allocation
        self.incrementSize = self.maxAllocated

    def getFirst(self):
        if self.numInList > 0:
            self.activeItem = 0
            return self.list[self.activeItem]
        else:
            self.activeItem = -1
            return None

    def getNext(self):
        if self.numInList > 0:
            self.activeItem += 1
            if self.activeItem > self.numInList:
                self.activeItem = self.numInList
                return None
            else:
                return self.list[self.activeItem]
        else:
            self.activeItem = -1
            return None

    def getActive(self):
        if self.activeItem > 0 and self.activeItem <= self.numInList:
            return self.get(self.activeItem)
        else:
            return None

    def setNew(self, value):
        self.add(value)

    def clear(self):
        self.activeItem = -1
        self.numInList = 0

    def add(self, p):
        """Returns index of item"""
        self.numInList += 1
        if self.numInList > self.maxAllocated:
            self.maxAllocated = self.maxAllocated + self.incrementSize
            # FIXME: Resize array
            self.list = [None] * self.maxAllocated
        self.list[self.numInList] = p
        self.activeItem = self.numInList
        return self.numInList

    def get(self, i):
        if (i < 1) or (i > self.numInList):
            return None
        else:
            self.activeItem = i
            return self.list[i]

    def size(self):
        return self.numInList

    def getActiveIndex(self):
        return self.activeItem
